Implement "live" reporters + many mocha reporter counter-parts, including time measurement

felixSchl opened the pull request:
I am building a new runner type that uses purescript-pipes to pipe results directly to the reporter, but also collects the results itself for final evaluation. This allows to stream results to the console, rather than running all tests first and then printing all of it.

It's still work in progress, but the goal is to eventually replace the current `Runner.run` implementation transparently in order to allow reporters to do more cool stuff :)

### asciinema casts:

* [streaming console reporter](https://asciinema.org/a/9sbvxkxl5n9i793z0p7l19jmp)
* [streaming spec reporter](https://asciinema.org/a/0x0d0wuk352b5d4psvex5o2ex)
* [streaming dot reporter](https://asciinema.org/a/ds051uq8k8vfut9sz225ihnxe)
* [streaming TAP reporter](https://asciinema.org/a/86h9uji95kvdedk6xrierr6ka)

### things to sort out:
* [x] require a proper purescript-spec release (depends on purescript-mmorph release)
* [x] should we store what is considered "slow" on the reporter or on the runner? I implemented on the reporter as it is in mocha, but thinking about it, it feels like it should be interpreted by the runner instead. @owickstrom , what do you think?
* [ ] i have implemented a [default summary](https://github.com/felixSchl/purescript-spec/blob/965bc0e20b732827b8fc7f4faeb279de92aad2e9/src/Test/Spec/Reporter/Base.purs#L92-L129) that looks like mocha's default summary report. But in order to keep things "as they were", I have left the current [console reporter's code in tact](https://github.com/felixSchl/purescript-spec/blob/965bc0e20b732827b8fc7f4faeb279de92aad2e9/src/Test/Spec/Reporter/Console.purs#L46-L72). We should decide what we want for a default summary and use it across at least all reporters currently implemented (I believe neither of them warrant a different summary look).
* [x] to get timeouts into the runner itself, it requires the `AVAR` and `TIMER` effects. This means that when defining a `Spec`, this won't work anymore `forall r. Spec r Unit`, as `r` must be `(avar :: AVAR, timer :: TIMER | r)`. Should we just `unsafeInterleaveEff` around this?

owickstrom commented:
This looks really neat! :) Will look at it in more detail.

felixSchl commented:
Thanks. I think my latest commit is moving in the right direction, esp around https://github.com/owickstrom/purescript-spec/pull/32/files#diff-edf85d78ba17883b458ee8575ec09337R33.

This way we can implement reporters simply by applying the `reporter` function to a custom `step` function (that receives events) and a custom `summarize` function (epilogue in mocha).

felixSchl commented:
Alright I have ported my work from #31 to using piped reporters. I will port the console reporter also, then move to the `Reporter` namespace. This new system gives us everything we had before plus the ability to report dynamically as the tests are running so we don't need both implementations, I think.

felixSchl commented:
I think this code is ready for review now. We won't be able to merge just yet because I am blocked on releasing a new version of purescript-pipes because of [Thimoteus/purescript-mmorph#2](https://github.com/Thimoteus/purescript-mmorph/pull/2). I am waiting for a new release of that library to depend on, then I can publish and update the dependency for this PR.

Also, it would be really helpful if you could take a look at https://github.com/felixSchl/purescript-pipes/blob/master/src/Pipes/Internal.purs#L141-L143 and https://github.com/felixSchl/purescript-pipes/blob/master/src/Pipes/ListT.purs#L67-L69 that would be much appreciated.

felixSchl commented:
I've added speed reporting to the bunch of new features as well.... I updated the asciinema casts if you wanted to have a look at it from a visual perspective

felixSchl commented:
I'll rebase this soon

felixSchl commented on src/Test/Spec.purs:
```diff
@@ -57,6 +60,13 @@ collect :: forall r. Spec r Unit
+countTests :: forall r. Spec r Unit -> Int
+countTests spec = execState (for (collect spec) go) 0
+  where
+  go (Describe _ _ xs) = void $ for xs go
```
should use `for_` to avoid array construction

owickstrom commented:
OK, I've finally gotten around to look at this in greater detail. It looks so promising, really taking purescript-spec to the next level, Felix! Great job, so far! :+1: 

About where to decide "what is slow", I agree that it should be a responsibility of the runner.

Regarding the formatting of a summary for all reporters, which one do you think is best? The legacy Console reporter was probably inspired by Hspec, but not an exact clone in formatting.

I'll look at the diffs as well, have to get a better understanding of purescript-pipes before I can make sense of it. :smile:

owickstrom commented:
You are taking over this knowledge map, step by step: https://codescene.io/projects/124/jobs/346/results/social/knowledge/by-author :grin:

felixSchl commented:
I implemented the TAP reporter. I will move the "what is slow" stuff into the runner next. I implemented the mocha style summary simply because I use that at work for all projects and I am used to it.

felixSchl commented:
Alright, so the `run` function can be configured now but also provides a default config. I've added timeouts as well (borrowed in part from purescript-test-unit, not sure how to credit them).

felixSchl commented:
Alright, I'll leave it at that for now, waiting for your feedback regarding the effect row stuff (I updated the todo check list in the summary).

owickstrom commented:
Cool! Can't say I found any single thing to remark on. :) Looks really good!

About the effect row, the sad thing is that it will break backwards compatibility for users on 0.10.x. It's not a very big change, so maybe we could be very clear about what needs to be done. Correct me if I'm wrong, but it should only be the effect row of specs?

felixSchl commented:
Yeah previously the effects caused by a spec where entirely driven by what the user wants to do with them, but now the spec runner causes some effects itself, so it will need to declare this. To keep backwards comp, we could mask it using `unsafeInterleaveEff`

owickstrom commented:
@felixSchl yeah, but that makes me a bit uneasy as well. :disappointed: Don't really know the conventions, but could using a custom effect be a middle ground? So we can hide whatever we have behind that, like `Spec (test :: TEST) Unit`. Then we'd at least not have to break backwards compatibility each time we change the implementation. Or we could encourage users to use a wildcard, like in your pull request. What do you think?

felixSchl commented:
Alright, I took care of it, but in the end I overshot a little: I also removed the return type for `Spec` since it was not used anywhere and there's not a single function that would suggest it makes sense to keep it. But I can back paddle on that if you want as it does break backwards compat to a small degree (users would have to update from `mySpec :: forall e. Spec e Unit` to `mySpec :: forall e. Spec e`.

owickstrom commented:
Yes, I think we should keep the return type. People may have utility functions for providing values to a spec, say a setup function that has type `Spec e MyState` which they need to be monadic. Plus, as you said, keeping it would mean one less breakage point.

felixSchl commented:
Good point, will back up the bus tomorrow morning

owickstrom commented:
Cool, thanks!

felixSchl commented:
I forced pushed the reversal, now the latest commit only contains the effect row stuff

owickstrom commented:
OK, I will have look as soon as possible. :)

owickstrom commented:
It seems the `purescript-pipes` dependency is missing from `bower.json`. I tried installing it manually and got a bunch of version conflicts, and the version of pipes is `0.0.0` which seemed a bit strange(?).

felixSchl commented:
I am blocked on releasing an update because of this: https://github.com/Thimoteus/purescript-mmorph/pull/2. I could, of course, depend on my github fork's sha in purescript-pipes, but doesn't feel right. That said, I also send thimoteus an out of channel email w/o response.

owickstrom commented:
We should get that properly released first. No updates? :/

felixSchl commented:
No updates. [Thimoteus](https://github.com/Thimoteus) appears to have gone off the radar. I wonder what's the right way to deal with this scenario.

owickstrom commented:
How long have you waited? Maybe for a (temporary) fork of the dependency. Not sure how that works with Bower publishing though?

felixSchl commented:
I have opened the PR to purescript-mmorph on 6th of November 2016. I have low faith it will get merged anytime soon because the owner seems to have vanished from the face of the earth. We could easily release a version of purescript-pipes that points to my fork.

felixSchl commented:
I have checked with the guys on IRC, depending on the fork is the only option at this point, but @natefaubion suggested purescript-mmorph has contrib potential, where it would see active maintenance (I volunteered until the original maintainer resurfaces). But for now, as far as purescript-spec is concerned, all should be working now.

owickstrom commented:
Yeah, I checked the IRC logs, and it seems reasonable. I'll try around a bit with this PR now, and then I think we should merge if nothing else comes up. brb :)

owickstrom commented:
Found that the colors from the `consoleReporter` does not work very well with some dark color schemes (iTerm 2):

![image](https://cloud.githubusercontent.com/assets/1464328/21860055/3e8c9712-d82f-11e6-8aa5-f539eb07747c.png)

I found the `Colors` module, where Pass maybe should be the default text color of the terminal instead? What do you think?

owickstrom commented:
Also, I found some minor warnings, all right if I fix those?

felixSchl commented:
Yes, please do. Do I give write rights to my fork or what?

owickstrom commented:
Yes, it seems so. :)

owickstrom commented:
Should I push the changes there?

felixSchl commented:
Entirely up to you, I gave you push rights on that repo if you want to finalize things before merging. If you prefer to just merge and then fix up, that's cool with me too

owickstrom commented:
Yeah, I'll push so that you can have a look at my changes too. :)

owickstrom commented:
Mostly unused imports and such, now fixed. If you don't have anything else, I think it's ready to be merged.

I'll update docs and tweet about it once it's on master. :)

felixSchl commented on src/Test/Spec/Runner.purs:
```diff
@@ -177,11 +173,11 @@ run' config reporters spec = void do
                      Process.exit 1
 
-    onSuccess :: Array (Group Result) -> Eff _ Unit
+    onSuccess :: Array (Group Result) -> Eff (SpecEffects e) Unit
```
Oh I must have missed those

owickstrom replied:
No problemos, should all be fixed now.

owickstrom replied:
I just pushed a change to use `SpecEffect` i the example also.

felixSchl commented:
Looked over it, looks good! I am happy for it to be merged

felixSchl commented:
what about the color issue though?

owickstrom commented:
Oh, right. Give me a minute, I'll try changing the Pass color to a default color.

owickstrom commented:
Exact same issue discussed around Mocha here: https://github.com/mochajs/mocha/issues/802. They concluded that it was a theme issue, not a problem with using a gray color. What do you say? Leave it as is? I'm not 100% sure, but I think I agree that themes should display gray properly, and we should not have to code around Solarized issues.

owickstrom commented:
Should have tried this before, but anyway... I ran this branch together with purescript-spec-discovery and found some issues regard effect rows. They're fixed now by using SpecEffects basically everywhere, also for user Aff values (Spec bodies) and Eff values (the resulting program when run).

felixSchl commented:
Re mocha colors: I'd go with the verdict reached in mochajs/mocha#802. Re `SpecEffects`, I guess it makes sense to only introduce the effects when needed, i.e. when `run`ning the spec.

felixSchl commented:
Funny you renamed it to `RunnerEffects`, I was thinking to suggest renaming it sth similar

owickstrom commented:
Yeah, seemed like a good name. :) I had to split out PROCESS to make the Mocha runner work in the browser. We should probably split out the PROCESS-dependent runner included in purescript-spec, along with the CONSOLE-dependent reporters. We can of course do this later, and it will be a breaking change, but I've seen some confusion around what is NodeJS/Browser-specific in purescript-spec. We could have a purescript-spec-node package perhaps.
